### PR TITLE
Decouple scheduled task execution from task scheduling.

### PR DESCRIPTION
**Pull request by chrisvest**
By making the scheduled executor service send the tasks to be executed in their respective pools, instead of the scheduling pool, and instead use the scheduling pool exclusively to trigger said tasks.

This also includes a handful of refactorings around the job scheduling infrastructure. And all threads started by the `CentralJobScheduler` now run within `ThreadGroups` derived from their scheduling `Group`s.

### CODE REVIEW

File: enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/PageCacheWarmerKernelExtension.java [hunk 1]
```diff
@@ -126,13 +126,7 @@ private void tryStartProfile()
     {
         if ( profilingStarted.compareAndSet( false, true ) )
         {
-            // At this point, we are currently executing inside a *scheduled* executor thread. There are, at this time
```
Comment (klaren) [inline enterprise/kernel/src/main/java/org/neo4j/kernel/impl/pagecache/PageCacheWarmerKernelExtension.java:129]: :tada:

File: community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/CentralJobScheduler.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+import org.neo4j.time.Clocks;
+
+import static org.neo4j.kernel.impl.util.DebugUtil.trackTest;
+
+public class CentralJobScheduler extends LifecycleAdapter implements JobScheduler
+{
+    private static final AtomicInteger INSTANCE_COUNTER = new AtomicInteger();
+    private static final Group SCHEDULER_GROUP = new Group( "Scheduler" );
+
+    private TimeBasedTaskScheduler scheduler;
```
Comment (klaren) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/CentralJobScheduler.java]: `final`, same for `schedulerThread`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/ThreadPoolManager.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        ThreadPool threadPool = getThreadPool( group );
+        return threadPool.submit( job );
+    }
+
+    public InterruptedException shutDownAll()
+    {
+        pools.forEach( ( group, pool ) -> pool.cancelAllJobs() );
+        pools.forEach( ( group, pool ) -> pool.shutDown() );
+        return pools.values().stream().reduce( null,
```
Comment (klaren) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/ThreadPoolManager.java]: I would prefer to have a simpler map-reduce instead:
```java
return pools.values().stream().map( ThreadPool::getShutdownException ).reduce( null, Exceptions::chain );
```

File: community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.LockSupport;
+
+import org.neo4j.scheduler.JobScheduler.Group;
+import org.neo4j.scheduler.JobScheduler.JobHandle;
+import org.neo4j.time.SystemNanoClock;
+
+final class TimeBasedTaskScheduler implements Runnable
```
Comment (klaren) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:31]: This class seems overly complicated for what needs to be done. A simple blocking priority queue would surely suffice.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/ScheduledTask.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.neo4j.concurrent.BinaryLatch;
+import org.neo4j.scheduler.JobScheduler;
+import org.neo4j.scheduler.JobScheduler.CancelListener;
+import org.neo4j.scheduler.JobScheduler.JobHandle;
+
+final class ScheduledTask implements JobHandle
```
Comment (klaren) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/ScheduledTask.java]: A `javadoc` that documents the states and transitions would be useful.

**Review by klaren — CHANGES_REQUESTED** [review]
This is a move in the right direction, nice to get some structure around the scheduler.

Since it feels like my requested changes will change a lot of the code, let's start with these so I don't review code that will go away anyway.

Was there any particular reason for not using `ScheduledThreadPoolExecutor` as a dispatcher for the other thread pools? Exactly the way you solved it for the profiler before.

**Comment by chrisvest** [comment]
> Was there any particular reason for not using ScheduledThreadPoolExecutor as a dispatcher for the other thread pools?

Yes, I tried this at first, but it became very unwieldy to trace cancellations and ExecutionExceptions back and forth, not to mention also handling overdue tasks and avoiding overlapping execution. The dedicated scheduler thread is definitely simpler.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java (hunk 4 above)
Comment (chrisvest) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:31]: You mean a DelayQueue? I don't think it's that complicated for balancing reaction time and proper shut down.

Comment (chrisvest) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:31]: I side-stepped the concurrency questions of the _blocking_ priority queue, and kept the existing concurrency scheme but switched to an ordinary priority queue.

**Comment by chrisvest** [comment]
@klaren I believe I've address all of your comments.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java (hunk 4 above)
Comment (klaren) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:31]: I was referring to the inbox and queue split as well. Can't submit just add it to the queue directly?

If we're having trouble with contention on the submit method, I'm more than sure we don't need to worry about timeliness.

Comment (chrisvest) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:31]: I like that they are split. They do different things as it is now. And one has to deal with multi-threading, while the other is single threaded and carefree.

Comment (klaren) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:31]: I don't see how multiple threads submitting and one thread polling would change that. The timekeeper thread will only work on the first object in the queue each time.

**Comment by chrisvest** [comment]
@klaren Take another look.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    private final ThreadPoolManager pools;
+    private final PriorityBlockingQueue<ScheduledJobHandle> delayedTasks;
+    private volatile Thread timeKeeper;
+    private volatile boolean stopped;
+
+    TimeBasedTaskScheduler( SystemNanoClock clock, ThreadPoolManager pools )
+    {
+        this.clock = clock;
+        this.pools = pools;
+        delayedTasks = new PriorityBlockingQueue<>( 42, DEADLINE_COMPARATOR );
```
Comment (klaren) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:47]: While 42 might be the answer to everything, somehow I doubt that includes optimal initial capacity for a blocking priority queue. But I don't what the answer to that would be...

**Review by klaren — APPROVED** [review]
I'm happy with the current state! One magic constant to ponder upon before merging.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java (hunk 6 above)
Comment (chrisvest) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/scheduler/TimeBasedTaskScheduler.java:47]: The queue grows automatically, and you have to choose something in order to provide a comparator, so here we are.

**Pull request merged**